File: focus.js
/* a module to allow elements to know when they have lost focus, */
/* that is, when a click occurs outside their bounds */

var focus = (function() {
	var focus = {};

	// the set of elements that want to be notified
	var registrants = $();
	var callbacks = {};

	focus.register = function(registrant, callback) {
		// store the callback
		$(registrant).data("clickaway", {callback: callback});

		// add registrant to the registry
		registrants = registrants.add(registrant);
	}

	// wait until dom is ready
	$(function() {
		// add click handler to body
		$("body").click(function(event) {
			// search registry
			registrants.each(function() {
				// if the event target is not a descendant of the registrant
				if(!$(this).has(event.target).length) {
					// notify the registrant of the click away
					$(this).data("clickaway").callback();
				}
			});
		});
	});

	return focus;

}());

File: Focus.dart
/** [Focus] is a class to allow elements to know when they
 *	have lost focus, that is, when a click occurs outside their bounds
 **/

class Focus {
  // The element to watch for loss of focus
  Element element;
  // The function to be called when focus is lost
  var callback;

  Focus(this.element, this.callback) {
    // register
    register(this);
  }

  // The list of Focus instances watching for focus loss
  final static List<Focus> registrants = new List<Focus>();
  // Register a new element
  static register(Focus focus) {
    registrants.add(focus);
  }
  // TODO unregister?
  // Initialize monitoring
  static Init() {
    // add click handler to document
    document.on.click.add((event) {
      // for each registrant,
      registrants.forEach((focus) {
        // if click is outside element, call callback
        if(!focus.element.contains(event.target)) registrant.callback();
      });
    });
  }
}